src/gs-update-monitor.c

```diff
@@ -150,6 +190,269 @@ no_updates_for_a_week (GsUpdateMonitor *monitor)
+	GMainContext *context = NULL;
+	MwscScheduler *scheduler = NULL;
+
+	if (monitor->scheduler != NULL)
+		return TRUE;
+
+	context = g_main_context_default ();
+
+	mwsc_scheduler_new_async (cancellable,
+				  (GAsyncReadyCallback) async_result_cb,
+				  &new_result);
+	while (new_result == NULL)
+		g_main_context_iteration (context, TRUE);
+
+	scheduler = mwsc_scheduler_new_finish (new_result, error);
+
+	if (scheduler == NULL)
+		return FALSE;
+
+	g_signal_connect_swapped (scheduler, "invalidated",
+				  (GCallback) monitor_clear_scheduler,
+				  monitor);
+	monitor->scheduler = scheduler;
+
+	return TRUE;
+}
+
+static void
+scheduled_entry_invalidated_cb (MwscScheduleEntry *entry,
+				const GError *error,
+				UpdateScheduleHelper *helper)
+
+{
+	const gchar *app_id = gs_app_get_unique_id (helper->app);
+
+	g_debug ("Removing scheduled update of app %s", app_id);
+	g_hash_table_remove (helper->monitor->scheduled_updates, app_id);
+}
+
+static void
+schedule_update (GsUpdateMonitor *monitor,
+		 GsApp *app,
+		 GCancellable *cancellable)
+{
+	g_auto(GVariantDict) parameters_dict = G_VARIANT_DICT_INIT (NULL);
+	g_autoptr(GVariant) parameters = NULL;
+	g_autoptr(GError) local_error = NULL;
+	g_autoptr(GAsyncResult) new_result = NULL;
+	UpdateScheduleHelper *helper = NULL;
+	GMainContext *context = NULL;
+	const gchar *app_id = gs_app_get_unique_id (app);
+
+	helper = g_hash_table_lookup (monitor->scheduled_updates, app_id);
+	if (helper != NULL) {
+		/* replace the app that's scheduled, in case the object is
+		 * different */
+		if (helper->app != app) {
+			GCancellable *app_cancellable = gs_app_get_cancellable (helper->app);
+			g_cancellable_cancel (app_cancellable);
+			g_set_object (&helper->app, app);
+		}
+
+		return;
+	}
+
+	g_variant_dict_insert (&parameters_dict, "resumable", "b", FALSE);
+	parameters = g_variant_ref_sink (g_variant_dict_end (&parameters_dict));
+
+	if (!setup_scheduler (monitor, cancellable, &local_error)) {
+		g_warning ("Could not schedule update for app %s: %s",
+			   app_id, local_error->message);
+		return;
+	}
+
+	if (helper == NULL) {
+		helper = g_new0 (UpdateScheduleHelper, 1);
+		helper->monitor = monitor;
+		helper->app = g_object_ref (app);
+	}
+
+	mwsc_scheduler_schedule_async (monitor->scheduler, parameters, cancellable,
+				       (GAsyncReadyCallback) async_result_cb,
+				       &new_result);
+
+	context = g_main_context_default ();
+	while (new_result == NULL)
+		g_main_context_iteration (context, TRUE);
+
+	helper->entry = mwsc_scheduler_schedule_finish (monitor->scheduler,
+							new_result, &local_error);
+
+	if (helper->entry == NULL) {
+		g_warning ("Failed to get schedule entry for updating app %s: %s",
+			   app_id, local_error->message);
+		return;
+	}
+
+	g_debug ("Scheduling new update for app %s with entry id %s", app_id,
+		 mwsc_schedule_entry_get_id (helper->entry));
+	g_hash_table_insert (monitor->scheduled_updates, g_strdup (app_id),
+			     g_steal_pointer (&helper));
```

Comment:
`helper` is a local variable and not a auto pointer to `UpdateScheduleHelper`, so you don't need to use `g_steal_pointer` here as it won't make any difference to set it to `NULL` before exiting the function's scope.

Reply:
It used to be an autoptr elsewhere... copy&paste mistake I think. Thanks for spotting.

Reply:
Ok